src/app/tests/TestClusterObjectUtils.cpp

```diff
@@ -0,0 +1,782 @@
+    static void TestClusterObjectUtils_EncAndDecIterableNestedStructList(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_EncAndDecIterableDoubleNestedStructList(nlTestSuite * apSuite, void * apContext);
+
+    static void TestClusterObjectUtils_OptionalFields(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_ExtraField(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_InvalidSimpleFieldTypes(nlTestSuite * apSuite, void * apContext);
+    static void TestClusterObjectUtils_InvalidListType(nlTestSuite * apSuite, void * apContext);
+
+private:
+    void SetupBuf();
+    void DumpBuf();
+    void SetupReader();
+
+    chip::System::TLVPacketBufferBackingStore mStore;
+    chip::TLV::TLVWriter mWriter;
+    chip::TLV::TLVReader mReader;
+    nlTestSuite *mpSuite;
+};
+
+using namespace chip::TLV;
+
+TestClusterObjectUtils gTestClusterObjectUtils;
+
+void TestClusterObjectUtils::SetupBuf()
+{
+    chip::System::PacketBufferHandle buf;
+   
+    buf = System::PacketBufferHandle::New(1024);
+    mStore.Init(std::move(buf));
+
+    mWriter.Init(mStore);
+    mReader.Init(mStore);
+}
+
+void TestClusterObjectUtils::DumpBuf()
+{
+    TLV::TLVReader reader;
+    reader.Init(mStore);
+
+    //
+    // TODO: Check-in TLV Text pretty printer in separate PR
+    //
+    //chip::TLV::Utilities::Print(reader);
+}
+
+void TestClusterObjectUtils::SetupReader()
+{
+    CHIP_ERROR err;
+
+    mReader.Init(mStore);
+    err = mReader.Next();
+
+    NL_TEST_ASSERT(mpSuite, err == CHIP_NO_ERROR);
+}
+
+template <typename T>
+struct TagValuePair {
+    uint64_t tag;
+    T& value;
+};
+
+template <typename T>
+TagValuePair<T> MakeTagValuePair(uint64_t tag, T &value) {
+    return TagValuePair<T>{tag, value};
+}
+
+template<typename... ArgTypes> CHIP_ERROR EncodeStruct(TLV::TLVWriter &writer, uint64_t tag, ArgTypes... Args)
+{
+  using expand_type = int[];
+  TLV::TLVType type;
+
+  ReturnErrorOnFailure(writer.StartContainer(tag, TLV::kTLVType_Structure, type));
+  (void)(expand_type{ (ClusterObjectUtils::Encode(writer, Args.tag, Args.value), 0)... });
+  ReturnErrorOnFailure(writer.EndContainer(type));
+
+  return CHIP_NO_ERROR;
+}
+
+void TestClusterObjectUtils::TestClusterObjectUtils_EncAndDecSimpleStruct(nlTestSuite * apSuite, void * apContext)
+{
+    CHIP_ERROR err;
+    chip::app::TestClusterObjectUtils *_this = static_cast<chip::app::TestClusterObjectUtils *>(apContext);
+
+    _this->mpSuite = apSuite;
+    _this->SetupBuf();
+
+    //
+    // Encode
+    //
+    {
+        chip::app::clusters::TestCluster::SimpleStruct::Type t;
+        uint8_t buf[4] = {0, 1, 2, 3};
+        char strbuf[10] = "chip";
+        
+        t.a = 20;
+        t.b = true;
+        t.c = chip::app::clusters::TestCluster::SimpleEnum::VALUEA;
+        t.d = buf;
+
+        // TODO: Bug in CHIPTLVWriter/Reader that don't quite deal with the null-terminator correctly.
```

Comment:
What is this bug, exactly?  This is not enough description to figure out what is meant.

Reply:
For now, I've decided to just remove the TODO, and keep the line as is.